File: fsharp-app/Example3_4.fs
module Example3_4

open System

type FundHolding = {
    SecurityId: string
    Units: decimal
    MarketValue: decimal
    TargetWeight: decimal
}

type DriftResult = {
    Holding: FundHolding
    CurrentWeight: decimal
    Drift: decimal
}

type SizedTrade = {
    DriftResult: DriftResult
    RequiredTradeValue: decimal
}

type ConstrainedOrder = {
    SizedTrade: SizedTrade
    AdjustedTradeValue: decimal
    Action: string
}

let calculateDrift (totalNav: decimal) (holding: FundHolding) : DriftResult =
    let currentWeight = holding.MarketValue / totalNav
    { Holding = holding; CurrentWeight = currentWeight; Drift = currentWeight - holding.TargetWeight }

let sizeRequiredTrade (totalNav: decimal) (drift: DriftResult) : SizedTrade =
    { DriftResult = drift; RequiredTradeValue = -drift.Drift * totalNav }

let applyConstraints
    (minTradeSize: decimal)
    (availableCash: decimal)
    (trades: SizedTrade list)
    : ConstrainedOrder list =
    let sells, buys = trades |> List.partition (fun t -> t.RequiredTradeValue < 0m)
    let constrainedSells =
        sells |> List.map (fun t ->
            if abs t.RequiredTradeValue < minTradeSize then
                { SizedTrade = t; AdjustedTradeValue = 0m; Action = "SKIP" }
            else
                { SizedTrade = t; AdjustedTradeValue = t.RequiredTradeValue; Action = "SELL" })
    let totalSellProceeds = constrainedSells |> List.sumBy (fun o -> abs o.AdjustedTradeValue)
    let totalAvailableCash = availableCash + totalSellProceeds
    let totalBuyDemand = buys |> List.sumBy (fun t -> t.RequiredTradeValue)
    let constrainedBuys =
        buys |> List.map (fun t ->
            if abs t.RequiredTradeValue < minTradeSize then
                { SizedTrade = t; AdjustedTradeValue = 0m; Action = "SKIP" }
            elif totalBuyDemand > totalAvailableCash then
                let scale = totalAvailableCash / totalBuyDemand
                { SizedTrade = t; AdjustedTradeValue = t.RequiredTradeValue * scale; Action = "BUY (scaled)" }
            else
                { SizedTrade = t; AdjustedTradeValue = t.RequiredTradeValue; Action = "BUY" })
    constrainedSells @ constrainedBuys


let runExample3_4 =
    // --- Tests: each stage tested independently with minimal data ---
    let test (name: string) (expected: 'a) (actual: 'a) =
        let status = if expected = actual then "PASS" else "FAIL"
        printfn $"  [{status}] {name} (expected: {expected}, got: {actual})"

    // Test drift calculation — needs only a decimal and a FundHolding
    printfn "Drift calculation tests:"
    let overweightHolding = { SecurityId = "BOND-ETF"; Units = 0m; MarketValue = 400_000m; TargetWeight = 0.30m }
    let drift1 = calculateDrift 1_000_000m overweightHolding
    test "overweight current weight" 0.40m drift1.CurrentWeight
    test "overweight drift is +10%" 0.10m drift1.Drift

    let underweightHolding = { SecurityId = "EQ-ETF"; Units = 0m; MarketValue = 200_000m; TargetWeight = 0.30m }
    let drift2 = calculateDrift 1_000_000m underweightHolding
    test "underweight drift is -10%" -0.10m drift2.Drift

    let onTargetHolding = { SecurityId = "INTL-ETF"; Units = 0m; MarketValue = 300_000m; TargetWeight = 0.30m }
    let drift3 = calculateDrift 1_000_000m onTargetHolding
    test "on-target drift is 0%" 0.00m drift3.Drift

    // Test trade sizing — needs only a decimal and a DriftResult
    printfn "\nTrade sizing tests:"
    let sellDrift = { Holding = overweightHolding; CurrentWeight = 0.40m; Drift = 0.10m }
    let sellTrade = sizeRequiredTrade 1_000_000m sellDrift
    test "overweight produces sell" -100_000m sellTrade.RequiredTradeValue

    let buyDrift = { Holding = underweightHolding; CurrentWeight = 0.20m; Drift = -0.10m }
    let buyTrade = sizeRequiredTrade 1_000_000m buyDrift
    test "underweight produces buy" 100_000m buyTrade.RequiredTradeValue

    // Test constraints — needs a list of SizedTrades, no Fund object at all
    printfn "\nConstraint tests:"

    // Helper to make a SizedTrade quickly for testing
    let makeTrade (secId: string) (reqValue: decimal) : SizedTrade =
        let h = { SecurityId = secId; Units = 0m; MarketValue = 0m; TargetWeight = 0m }
        let d = { Holding = h; CurrentWeight = 0m; Drift = 0m }
        { DriftResult = d; RequiredTradeValue = reqValue }

    let orders1 = applyConstraints 5_000m 40_000m [
        makeTrade "SELL-A" -80_000m
        makeTrade "BUY-B" 200_000m
        makeTrade "TOO-SMALL" 2_000m
    ]

    test "small trade skipped" "SKIP" (orders1 |> List.find (fun o -> o.SizedTrade.DriftResult.Holding.SecurityId = "TOO-SMALL")).Action
    test "sell passes through" -80_000m (orders1 |> List.find (fun o -> o.SizedTrade.DriftResult.Holding.SecurityId = "SELL-A")).AdjustedTradeValue

    // Available = 40K + 80K sell = 120K. Buy wants 200K. Scale = 120/200 = 0.6
    let buyOrder = orders1 |> List.find (fun o -> o.SizedTrade.DriftResult.Holding.SecurityId = "BUY-B")
    test "buy scaled to available cash" 120_000m buyOrder.AdjustedTradeValue

    // Order independence — same trades, reversed input list
    let orders2 = applyConstraints 5_000m 40_000m [
        makeTrade "TOO-SMALL" 2_000m
        makeTrade "BUY-B" 200_000m
        makeTrade "SELL-A" -80_000m
    ]
    let buyOrder2 = orders2 |> List.find (fun o -> o.SizedTrade.DriftResult.Holding.SecurityId = "BUY-B")
    test "order-independent: same result reversed" buyOrder.AdjustedTradeValue buyOrder2.AdjustedTradeValue

// 14 tests. No Fund object needed for any of them. No pre-setting internal fields.
// Each function tested with just its actual inputs.
// Constraint test proves order independence — impossible to verify in the C# version.

(* Why F# prevents it: Each stage is a pure function with a clear signature. Testing drift needs a decimal
 and a FundHolding — not a fully constructed Fund with irrelevant fields. Testing constraints needs a SizedTrade 
 list — not a Fund with pre-mutated RequiredTradeValue fields. No test depends on list ordering because the logic 
 is order-independent by construction. And you can trivially prove that with a test — something the C# version can 
 never pass because its mutation makes ordering a hidden input. *)

File: csharp-app/Example3_4.cs
namespace Example3_4;

using System;
using System.Collections.Generic;
using System.Linq;


public class FundHolding
{
    public string SecurityId { get; set; } = "";
    public decimal Units { get; set; }
    public decimal MarketValue { get; set; }
    public decimal CurrentWeight { get; set; }
    public decimal TargetWeight { get; set; }
    public decimal Drift { get; set; }
    public decimal RequiredTradeValue { get; set; }
    public decimal AdjustedTradeValue { get; set; }
}

public class Fund
{
    public string FundId { get; set; } = "";
    public List<FundHolding> Holdings { get; set; } = new();
    public decimal TotalNav { get; set; }
    public decimal CashBalance { get; set; }
    public decimal MinCashReserve { get; set; }
    public decimal MinTradeSize { get; set; }
}

public class RebalanceEngine
{
    private Fund? _fund;

    public void LoadFund(Fund fund) => _fund = fund;

    public void CalculateDrift()
    {
        foreach (var h in _fund!.Holdings)
        {
            h.CurrentWeight = h.MarketValue / _fund.TotalNav;
            h.Drift = h.CurrentWeight - h.TargetWeight;
        }
    }

    public void SizeRequiredTrades()
    {
        foreach (var h in _fund!.Holdings)
        {
            h.RequiredTradeValue = -h.Drift * _fund.TotalNav;
        }
    }

    public void ApplyConstraints()
    {
        var totalSellValue = _fund!.Holdings
            .Where(h => h.RequiredTradeValue < 0)
            .Sum(h => h.RequiredTradeValue);

        foreach (var h in _fund.Holdings)
        {
            if (Math.Abs(h.RequiredTradeValue) < _fund.MinTradeSize)
            {
                h.AdjustedTradeValue = 0;
                continue;
            }
            var availableCash = _fund.CashBalance + Math.Abs(totalSellValue) - _fund.MinCashReserve;
            if (h.RequiredTradeValue > 0 && h.RequiredTradeValue > availableCash)
            {
                h.AdjustedTradeValue = availableCash;
                _fund.CashBalance -= h.AdjustedTradeValue;
            }
            else
            {
                h.AdjustedTradeValue = h.RequiredTradeValue;
            }
        }
    }

    public void Rebalance()
    {
        CalculateDrift();
        SizeRequiredTrades();
        ApplyConstraints();
    }
}

public class Example
{
    // Helper to assert
    static void Assert(string name, bool condition)
    {
        Console.WriteLine($"  [{(condition ? "PASS" : "FAIL")}] {name}");
    }

    public static void Run()
    {
        Console.WriteLine("Test: Drift calculation for overweight holding");
        {
            // To test JUST drift, we still need a full Fund with a Holdings list,
            // because CalculateDrift reads _fund.TotalNav and mutates holdings in place.
            var fund = new Fund
            {
                FundId = "TEST",
                TotalNav = 1_000_000m,
                CashBalance = 0m,
                MinCashReserve = 0m,
                MinTradeSize = 0m,
                Holdings = new List<FundHolding>
                {
                    new()
                    {
                        SecurityId = "BOND-ETF",
                        MarketValue = 400_000m,
                        TargetWeight = 0.30m
                    }
                }
            };

            var engine = new RebalanceEngine();
            engine.LoadFund(fund);
            engine.CalculateDrift();

            // Test the mutated holding
            Assert("current weight is 40%", fund.Holdings[0].CurrentWeight == 0.40m);
            Assert("drift is +10%", fund.Holdings[0].Drift == 0.10m);
        }

        Console.WriteLine("\nTest: Trade sizing");
        {
            // To test sizing, we need drift to be already set on the holding.
            // But SizeRequiredTrades reads h.Drift — which CalculateDrift sets.
            // So we either: (a) call CalculateDrift first (testing two stages),
            // or (b) manually set Drift on the holding (fragile setup that
            // bypasses the actual pipeline).
            var fund = new Fund
            {
                FundId = "TEST",
                TotalNav = 1_000_000m,
                CashBalance = 0m,
                MinCashReserve = 0m,
                MinTradeSize = 0m,
                Holdings = new List<FundHolding>
                {
                    new()
                    {
                        SecurityId = "BOND-ETF",
                        MarketValue = 400_000m,
                        TargetWeight = 0.30m,
                        Drift = 0.10m  // manually pre-set — skipping CalculateDrift
                    }
                }
            };

            var engine = new RebalanceEngine();
            engine.LoadFund(fund);
            engine.SizeRequiredTrades();

            // Drift of +10% on 1M NAV => sell 100K
            Assert("required trade is -100K (sell)", fund.Holdings[0].RequiredTradeValue == -100_000m);
        }

        Console.WriteLine("\nTest: Constraint application");
        {
            // To test constraints alone, we need RequiredTradeValue pre-set.
            // But we also need CashBalance, MinCashReserve, MinTradeSize on Fund.
            // And the result depends on Holdings list ORDER because of the
            // CashBalance mutation in the loop.
            var fund = new Fund
            {
                FundId = "TEST",
                TotalNav = 1_000_000m,
                CashBalance = 50_000m,
                MinCashReserve = 10_000m,
                MinTradeSize = 5_000m,
                Holdings = new List<FundHolding>
                {
                    new() { SecurityId = "SELL-THIS", RequiredTradeValue = -80_000m },
                    new() { SecurityId = "BUY-THIS",  RequiredTradeValue = 200_000m },
                    new() { SecurityId = "TOO-SMALL", RequiredTradeValue = 2_000m },
                }
            };

            var engine = new RebalanceEngine();
            engine.LoadFund(fund);
            engine.ApplyConstraints();

            Assert("small trade skipped", fund.Holdings[2].AdjustedTradeValue == 0m);
            Assert("sell passes through", fund.Holdings[0].AdjustedTradeValue == -80_000m);

            // What SHOULD BUY-THIS be?
            // Available = 50K cash + 80K sell proceeds - 10K reserve = 120K
            // Requested = 200K, so constrained to 120K
            Assert("buy constrained to available cash", fund.Holdings[1].AdjustedTradeValue == 120_000m);

            // But this test is ORDER-DEPENDENT. If BUY-THIS was first in the list
            // and SELL-THIS was second, the CashBalance mutation would produce
            // a different result. The test only passes for THIS ordering.
        }

        // Summary: each test required building a full Fund object with fields
        // that aren't relevant to the stage being tested. Testing constraints
        // alone required manually pre-setting RequiredTradeValue — which means
        // the test is coupled to internal implementation details of how data
        // flows between stages. And the constraint test is fragile because
        // it depends on list ordering.
    }
}